swiperefresh/src/main/java/com/google/accompanist/swiperefresh/SwipeRefresh.kt

```diff
@@ -0,0 +1,336 @@
+            // trigger
+            state.isRefreshing -> max(drag, -(state.indicatorOffset - state.refreshTrigger))
+            else -> max(drag, -state.indicatorOffset)
+        }
+        return if (distanceAvailable.absoluteValue > 0.5f) {
+            coroutineScope.launch {
+                state.dispatchRawDelta(distanceAvailable)
+            }
+            // Consume the consumed Y
+            Offset(x = 0f, y = distanceAvailable / DragMultiplier)
+        } else {
+            Offset.Zero
+        }
+    }
+
+    override suspend fun onPreFling(available: Velocity): Velocity = when {
+        // If we're currently refreshing, just animate back to the resting position
+        state.isRefreshing -> {
+            coroutineScope.launch {
+                state.animateOffsetTo(state.refreshTrigger)
+            }
+            // Don't consume any velocity, to allow the scrolling layout to fling
+            Velocity.Zero
+        }
+        // If we're dragging and scrolled past the trigger point, refresh!
+        state.isSwipeInProgress && state.indicatorOffset >= state.refreshTrigger -> {
+            onRefresh()
+            // Don't consume any velocity, to allow the scrolling layout to fling
+            Velocity.Zero
+        }
+        else -> Velocity.Zero
+    }.also {
+        // Reset the drag in progress state
+        state.isSwipeInProgress = false
+    }
+}
+
+/**
+ * A layout which implements the swipe-to-refresh pattern.
+ *
+ * The layout can be used whenever the user has the ability to refresh content via a vertical
+ * swipe gesture.
+ *
+ * Apps should provide a [onRefresh] block to be notified each time a swipe to refresh gesture
+ * is completed. That block is responsible for updating the [state] as appropriately,
+ * typically by setting [SwipeRefreshState.isRefreshing] to `true` once a 'refresh' has been
+ * started. Once a refresh has completed, the app should then set
+ * [SwipeRefreshState.isRefreshing] to `false`.
+ *
+ * If an app wishes to show just the progress animation, outside of a swipe refresh, it can
+ * set [SwipeRefreshState.isRefreshing] as required.
+ *
+ * @sample com.google.accompanist.sample.swiperefresh.SwipeRefreshSample
+ *
+ * @param state the state object to be used to control or observe the [SwipeRefresh] state.
+ * @param onRefresh Lambda which is invoked when a swipe to refresh gesture is completed.
+ * @param modifier the modifier to apply to this layout.
+ * @param swipeEnabled Whether the the layout should react to swipe gestures or not.
+ * @param indicator the indicator that represents the current state. By default this
+ * will be a [SwipeRefreshIndicator].
+ * @param content The content containing a scroll composable.
+ */
+@Composable
+fun SwipeRefresh(
+    state: SwipeRefreshState,
+    onRefresh: () -> Unit,
+    modifier: Modifier = Modifier,
+    swipeEnabled: Boolean = true,
+    indicator: @Composable SwipeRefreshIndicatorScope.() -> Unit = {
```

Comment:
What do you think if you just pass State as a parameter for this slot? Or, alternatively, just do nothing and make a SwipeRefreshIndicator to accept a state from the SwipeRefresh itself?

I don't see a big usability reason to have this scope in general, but maybe I'm missing smth?

Reply:
As discussed, done.